cargo-nextest/src/dispatch.rs

```diff
@@ -963,6 +971,68 @@ impl TestReporterOpts {
     }
 }
 
+#[derive(Debug, Clone, Copy)]
+struct TestOutputDisplayStreamsOpt {
+    stdout: Option<TestOutputDisplayOpt>,
+    stderr: Option<TestOutputDisplayOpt>,
+}
+
+impl FromStr for TestOutputDisplayStreamsOpt {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // expected input has three forms
+        // - "{value}":                         where value is one of [immediate, immediate-final, final, never]
+        // - "{stream}={value}":                where {stream} is one of [stdout, stderr]
+        // - "{stream}={value},{stream=value}": where the two {stream} keys cannot be the same
+        let (stdout, stderr) = if let Some((left, right)) = s.split_once(',') {
+            // the "{stream}={value},{stream=value}" case
+            let left = left
+                .split_once('=')
+                .map(|l| (l.0, TestOutputDisplayOpt::from_str(l.1, false)));
+            let right = right
+                .split_once('=')
+                .map(|r| (r.0, TestOutputDisplayOpt::from_str(r.1, false)));
+            match (left, right) {
+                (Some(("stderr", Ok(stderr))), Some(("stdout", Ok(stdout)))) => (Some(stdout), Some(stderr)),
+                (Some(("stdout", Ok(stdout))), Some(("stderr", Ok(stderr)))) => (Some(stdout), Some(stderr)),
+                (Some((stream @ "stdout" | stream @ "stderr", Err(_))), _) => return Err(format!("\n  unrecognized setting for {stream}: [possible values: immediate, immediate-final, final, never]")),
+                (_, Some((stream @ "stdout" | stream @ "stderr", Err(_)))) => return Err(format!("\n  unrecognized setting for {stream}: [possible values: immediate, immediate-final, final, never]")),
+                (Some(("stdout", _)), Some(("stdout", _))) => return Err("\n  stdout specified twice".to_string()),
+                (Some(("stderr", _)), Some(("stderr", _))) => return Err("\n  stderr specified twice".to_string()),
+                (Some((stream, _)), Some(("stdout" | "stderr", _))) => return Err(format!("\n  unrecognized output stream '{stream}': [possible values: stdout, stderr]")),
+                (Some(("stdout" | "stderr", _)), Some((stream, _))) => return Err(format!("\n  unrecognized output stream '{stream}': [possible values: stdout, stderr]")),
+                (_, _) => return Err("\n  [possible values: immediate, immediate-final, final, never], or specify one or both output streams: stdout={}, stderr={}, stdout={},stderr={}".to_string()),
+            }
```

Comment:
Hmm, can this be shortened? This feels excessive.

I'd probably do something like:

* split with `,` if possible, getting 1 or 2 entries from this (more than 2 should error out)
* if one entry, try parsing via `<stream>=<setting>` or then just `<setting>`
* if two entries, always try parsing as `<stream>=<setting>`, then at the end ensure that the streams are different

Reply:
It can be reduced to the two first lines and the last line, but that would result in significantly less helpful error messages.
I also don't see a way to reduce it without degrading the error messaging.

I could do a priority inversion by first checking if the value is a valid TestOutputDisplayOpt, which should be the most common case. And only then dealing with commas and equal signs.

Reply:
Yeah, trying to parse it first that way makes sense.